### DISCUSSION TITLE
How to print all request headers?

### DISCUSSION THREAD

**Discussion by elliotwaite** [litestar-org/litestar#2880] [Q&A] [answered]
When I run this app to try to print all the request headers:
```
from litestar import Litestar, get
from litestar.datastructures import Headers


@get("/", sync_to_thread=False)
def index(headers: Headers) -> str:
    print("headers:", headers.to_header_list())
    return "Hello, world!"


app = Litestar([index])
```
... I get the following error:
```
{"status_code":400,"detail":"Validation failed for GET http://127.0.0.1:8000/","extra":[{"message":"Unsupported type: <class 'dict'>","key":"headers"}]}
```

I feel like I'm misunderstanding something. Can anyone point out what I'm doing wrong?

Also, the [Default signature namespace](https://docs.litestar.dev/2/usage/routing/handlers.html#default-signature-namespace) section of the docs mentions that I can place the `Headers` import inside an `if TYPE_CHECKING:` block. But when I do this:

```
from typing import TYPE_CHECKING

from litestar import Litestar, get

if TYPE_CHECKING:
    from litestar.datastructures import Headers


@get("/", sync_to_thread=False)
def index(headers: Headers) -> str:
    print("headers:", headers.to_header_list())
    return "Hello, world!"


app = Litestar([index])
```
... I get the following error when I try to run the app:
```
  ...
  File "/.../app.py", line 10, in <module>
    def index(headers: Headers) -> str:
                       ^^^^^^^
NameError: name 'Headers' is not defined
```

Am I doing something wrong here as well?

I'm using Litestar version 2.4.3.

**Comment by FHU-yezi** [comment]
Try wrapping `Headers` with quotes?

**Reply by elliotwaite** [reply]
Thanks. Changing `header: Headers` to `header: "Headers"` fixed the second issue related to using the `if TYPE_CHECKING:` block.

However, I am still running into the first issue, related to the `Validation failed ...` error.

**Comment by Alc-Alc** [comment]
Does this do it for you?

```python
from litestar import Litestar, get, Request

@get("/", sync_to_thread=False)
def index(request: Request) -> str:
    print("headers:", request.headers.to_header_list())
    return "Hello, world!"

app = Litestar([index])
```

**Comment by provinzkraut** [comment, marked as answer]
Hi @elliotwaite, unfortunately, you've come across a bug that can't be fixed until version 3 because it would constitute a breaking change; The headers injected into the route handler via the `headers` parameter isn't actually a `Headers` instance but a `dict`. For now, if you need access to a `Headers` instance, you'll have to use the method @Alc-Alc provided. If you're fine with a dict, you can just type `headers` as a dict.

**Reply by elliotwaite** [reply]
Okay, got it. Thanks!

Both options worked for me:
* Using `request: Request` and `print("headers:", request.headers.to_header_list())`
* Or using `headers: dict` and `print("headers:", headers)`

I created a PR [here](https://github.com/litestar-org/litestar/pull/2882) to update the docs to mention that the type of `headers` is a `dict` instead of an instance of `Headers`.
